django/contrib/auth/management/__init__.py

```diff
@@ -4,14 +4,104 @@
+
+        # Build a dictionary allowing to do the correspondence between the
+        # current permission codenames and the new <codename, name> values.
+        current_to_new_perms = {}
+        for codename, name in _get_builtin_permissions(current_model._meta):
+            old_codename = re.sub(r'_%s$' % (new_model_lower), '_' + old_model_lower, codename)
+            if not backward:
+                current_to_new_perms[old_codename] = {'codename': codename, 'name': name}
+            else:
+                action = old_codename.split('_' + old_model_lower, 1)[0]
+                current_to_new_perms[codename] = {
+                    'codename': old_codename,
+                    'name': 'Can %s %s' % (action, camel_case_to_spaces(self.old_model)),
+                }
+
+        perms = Permission.objects.db_manager(db).filter(codename__in=current_to_new_perms).order_by('codename')
+        for perm in perms:
+            old_codename = perm.codename
+            perm.codename = current_to_new_perms[old_codename]['codename']
+            perm.name = current_to_new_perms[old_codename]['name']
+            perm.save(update_fields={'codename', 'name'})
+
+    def rename_forwards(self, apps, schema_editor):
+        self._rename(apps, schema_editor)
+
+    def rename_backwards(self, apps, schema_editor):
+        self._rename(apps, schema_editor, backward=True)
+
+
+def inject_rename_permissions_operations(plan=None, apps=global_apps, using=DEFAULT_DB_ALIAS, **kwargs):
+    """
+    Insert a `RenamePermissions` operation after every planned `RenameModel`
+    operation.
+    """
+    if plan is None:
+        return
+
+    # Determine whether or not the Permission model is available.
+    try:
+        Permission = apps.get_model('auth', 'Permission')
+    except LookupError:
+        available = False
+    else:
+        if not router.allow_migrate_model(using, Permission):
+            return
+        available = True
+
+    for migration, backward in plan:
+        if ((migration.app_label, migration.name) == ('auth', '0001_initial')):
+            # There's no point in going forward if the initial auth
+            # migration is unapplied as the Permission model will be
+            # unavailable from this point.
+            if backward:
+                break
+            else:
+                available = True
+                continue
+        # The Permission model is not available yet.
+        if not available:
+            continue
+        inserts = []
+        for index, operation in enumerate(migration.operations):
+            if isinstance(operation, migrations.RenameModel):
+                operation = RenamePermissions(
+                    migration.app_label, operation.old_name, operation.new_name
+                )
+                inserts.append((index + 1, operation))
+        for inserted, (index, operation) in enumerate(inserts):
+            migration.operations.insert(inserted + index, operation)
```

Comment:
Sorry for the delayed reply here. I'm still not a fan retrieving permission by verbose names at it seems likely to break.

Regarding the difficulty to detect whether or not `RenameContentType` was run already you should be able to take this decision at `RenamePermissions` insertion time by looking up if the operation following the `RenameModel` operation. I'd also be in favor of only injecting `RenamePermissions` operations after `RenameContentType` ones meaning no permission renaming will take place if `django.contrib.auth` is installed before `django.contrib.contenttypes` and even to raise a `RuntimeWarning` letting the user know no permission renaming took place because of this.

The ideal solution would be to add a check to make sure `contenttypes` is always installed before `auth` as it depends on it but that would require a larger discussion on the development mailing list as [most of the projects currently install these apps in the opposite order](https://github.com/django/django/blob/master/django/conf/project_template/project_name/settings.py-tpl#L35-L36).